File: backend_python/src/todo_api/infrastructure/repository/todo_repository.py
from sqlalchemy import delete, select
from sqlalchemy.orm import Session

from todo_api.domain.model.todo import Todo
from todo_api.domain.repository.context_provider import ContextProvider
from todo_api.domain.repository.errors import RepositoryNotFoundError
from todo_api.domain.repository.todo_repository import TodoRepository
from todo_api.infrastructure.repository.data_model.todo import TodoDataModel
from todo_api.utils.uuid import UUID7


class TodoRepositoryImpl(TodoRepository):
    def __init__(self, context_provider: ContextProvider[Session]) -> None:
        self.context_provider = context_provider

    def find_all(self) -> list[Todo]:
        stmt = select(TodoDataModel)
        session = self.context_provider.current()
        todos = session.scalars(stmt).all()

        return [todo.to_domain() for todo in todos]

    def find_by_id(self, todo_id: UUID7):
        stmt = select(TodoDataModel).where(TodoDataModel.id == str(todo_id))

        session = self.context_provider.current()
        todo_data_model = session.scalars(stmt).one_or_none()

        if todo_data_model is None:
            raise RepositoryNotFoundError(f"Todo with id {todo_id} not found")
        return todo_data_model.to_domain()

    def save(self, todo: Todo) -> None:
        session = self.context_provider.current()
        todo_data_model = TodoDataModel.from_domain(todo)

        session.merge(todo_data_model)
        session.flush()

    def delete(self, todo: Todo) -> None:
        session = self.context_provider.current()
        session.execute(delete(TodoDataModel).where(TodoDataModel.id == str(todo.id)))

File: backend_bun/src/infrastructure/repository/todoRepository.ts
import { Todo } from "@/domain/model/todo";
import type { ContextProvider } from "@/domain/repository/context";
import { RepositoryNotFoundError } from "@/domain/repository/errors";
import type { TodoRepository } from "@/domain/repository/todoRepository";
import type { Conn } from "./context";

type TodoDatamodel = {
  id: string;
  title: string;
  description: string | null;
  completed: 0 | 1;
};

function toTodoDatamodel(todo: Todo): TodoDatamodel {
  const { id, title, description, isCompleted } = todo.toPrimitives();
  return {
    id,
    title,
    description: description ?? null,
    completed: isCompleted ? 1 : 0,
  };
}

function fromTodoDatamodel(data: TodoDatamodel): Todo {
  return Todo.fromPrimitives({
    id: data.id,
    title: data.title,
    description: data.description ?? null,
    isCompleted: data.completed === 1,
  });
}

export class TodoRepositoryImpl implements TodoRepository {
  constructor(private readonly contextProvider: ContextProvider<Conn>) {}

  async save(todo: Todo): Promise<void> {
    const db = this.contextProvider.get();
    const data = toTodoDatamodel(todo);

    await db`INSERT INTO todos 
      (
        id, 
        title, 
        description, 
        completed
      ) 
    VALUES 
      (
        ${data.id}, 
        ${data.title}, 
        ${data.description}, 
        ${data.completed}
      ) 
    ON DUPLICATE KEY UPDATE 
      title = ${data.title}, 
      description = ${data.description}, 
      completed = ${data.completed}`;
  }

  async findById(id: string): Promise<Todo> {
    const db = this.contextProvider.get();
    const rows: TodoDatamodel[] =
      await db`SELECT * FROM todos WHERE id = ${id}`;
    if (rows.length === 0)
      throw new RepositoryNotFoundError(`Todo id(${id}) not found`);
    const todo = rows[0];
    return fromTodoDatamodel(todo);
  }

  async findAll(): Promise<Todo[]> {
    const db = this.contextProvider.get();
    const rows: TodoDatamodel[] =
      await db`SELECT id, title, description, completed FROM todos`;
    return rows.map((item) => fromTodoDatamodel(item));
  }

  async delete(todo: Todo): Promise<void> {
    const db = this.contextProvider.get();
    await db`DELETE FROM todos WHERE id = ${todo.id}`;
  }
}
